admin/server/usergroups.go

```diff
@@ -0,0 +1,696 @@
+	}
+
+	if org.AllUsergroupID != nil && group.ID == *org.AllUsergroupID {
+		return nil, status.Error(codes.InvalidArgument, "cannot add member to all-users group")
+	}
+
+	user, err := s.admin.DB.FindUserByEmail(ctx, req.Email)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	claims := auth.GetClaims(ctx)
+
+	orgRole, err := s.admin.DB.FindOrganizationMemberUsergroupRole(ctx, group.ID, org.ID)
+	if err != nil && !errors.Is(err, database.ErrNotFound) {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if orgRole != nil {
+		if !claims.OrganizationPermissions(ctx, org.ID).ManageOrgMembers {
+			return nil, status.Error(codes.PermissionDenied, "not allowed to add usergroup members")
+		}
+	}
+
+	projectRoles, err := s.admin.DB.FindProjectMemberUsergroupRoles(ctx, group.ID, org.ID)
+	if err != nil && !errors.Is(err, database.ErrNotFound) {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	for _, role := range projectRoles {
+		if !claims.ProjectPermissions(ctx, org.ID, role.ProjectID).ManageProjectMembers {
+			return nil, status.Error(codes.PermissionDenied, "not allowed to add usergroup members")
+		}
+	}
+
+	err = s.admin.DB.InsertUsergroupMemberUser(ctx, group.ID, user.ID)
+	if err != nil {
+		if errors.Is(err, database.ErrNotUnique) {
+			return nil, status.Error(codes.AlreadyExists, err.Error())
+		}
+		return nil, err
+	}
+
+	return &adminv1.AddUsergroupMemberUserResponse{}, nil
+}
+
+func (s *Server) ListUsergroupMemberUsers(ctx context.Context, req *adminv1.ListUsergroupMemberUsersRequest) (*adminv1.ListUsergroupMemberUsersResponse, error) {
+	observability.AddRequestAttributes(ctx,
+		attribute.String("args.org", req.Organization),
+		attribute.String("args.usergroup", req.Usergroup),
+	)
+
+	org, err := s.admin.DB.FindOrganizationByName(ctx, req.Organization)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	group, err := s.admin.DB.FindUsergroupByName(ctx, req.Usergroup, org.ID)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	claims := auth.GetClaims(ctx)
+
+	orgRole, err := s.admin.DB.FindOrganizationMemberUsergroupRole(ctx, group.ID, org.ID)
+	if err != nil && !errors.Is(err, database.ErrNotFound) {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if orgRole != nil {
+		if !claims.OrganizationPermissions(ctx, org.ID).ReadOrgMembers {
+			return nil, status.Error(codes.PermissionDenied, "not allowed to list usergroup members")
+		}
+	}
+
+	projectRoles, err := s.admin.DB.FindProjectMemberUsergroupRoles(ctx, group.ID, org.ID)
+	if err != nil && !errors.Is(err, database.ErrNotFound) {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	for _, role := range projectRoles {
+		if !claims.ProjectPermissions(ctx, org.ID, role.ProjectID).ReadProjectMembers {
+			return nil, status.Error(codes.PermissionDenied, "not allowed to list usergroup members")
+		}
+	}
```

Comment:
Might be missing something, but I think listing a group's members should just require `ReadOrgMembers` on the group's org?